docs/HowToFAQ/Errors-Resolver3.md

```diff
@@ -0,0 +1,678 @@
+module N { type T }
+
+import opened M
+import opened N
+const t: T
+
+```
+
+The stated type name has more than one declaration in scope, likely because both have been imported
+with `opened`. In that case the name must be qualified to indicate which declaration is intended.
+
+## **Error: Expected datatype: _type_**
+
+<!-- TODO -->
+
+## **Error: All elements of display must have some common supertype (got _type_, but needed type or type of previous elements is _type_)**
+
+<!-- TODO -->
+
+<!-- 3 instance -->
+
+## **Error: name of module (_name_) is used as a variable**
+
+```dafny
+module M {}
+const c := M
+```
+
+All names in Dafny (except for names of export sets) are in one namespace. Names in different 
+scopes can be distinguished by qualification. Names can also be redeclared in a nested scope
+with different properties. But if a name is visible, it must be used consistent with its
+declaration. So a name declared as a module cannot be used as a variable, even though it is 
+usually clear from context where module names are used and where expressions are used.
+
+
+<!-- 2 instances -->
+
+## **Error: name of type (_type_) is used as a variable**
+
+```dafny
+type t = int
+method m(i: int) {
+  t := i;
+}
+```
+
+All names in Dafny (except for names of export sets) are in one namespace. Names in different 
+scopes can be distinguished by qualification. Names can also be redeclared in a nested scope
+with different properties. But if a name is visible, it must be used consistent with its
+declaration. So a name declared as a type cannot be used as a variable, even though it is 
+usually clear from context where type names are used and where expressions are used.
+
+<!-- 2 instances -->
+
+## **Error: a two-state function can be used only in a two-state context**
+
+<!-- %no-check TODO - fix - incorrect error message for this example -->
+```dafny
+module M {
+  twostate function f(): int
+}
+const c := M.f()
+```
+
+A `twostate` function is a function of two different heaps.
+Accordingly it can be used only in situations in which there are two states
+in play (that is, a _two-state_ context).  One example is in a postcondition
+(_ensures_ clause) where the two states are the states at the beginning and end 
+of the method execution. Another two-state context is the body of a method, 
+where the states are the pre-state of the method and the current state at the
+location of the call. However, contexts outside of a method, such as initializations 
+of const declarations, are not two-state contexts.
+
+## **Error: a field must be selected via an object, not just a class name**
+
+<!-- TODO -->
+
+## **Error: member _name_ in type _type_ does not refer to a field or a function**
+
+<!-- TODO - not sure this is reachable -->
+
+## **Error: array selection requires an array_n_ (got _type_)**
+
+<!-- TODO - fix - different error message for this example -->
+
+```dafny
+const a: int
+const x: ORDINAL
+method m() 
+{  
+var c := a[0,0];
```

Comment:
Indentation would make this slightly nicer.

Reply:
Indented